src/lib/accessibility/FocusZone/FocusTrapZone.tsx

```diff
@@ -0,0 +1,322 @@
+import getUnhandledProps from '../../getUnhandledProps'
+import getElementType from '../../getElementType'
+import * as customPropTypes from '../../customPropTypes'
+
+export class FocusTrapZone extends React.Component<IFocusTrapZoneProps, {}>
+  implements IFocusTrapZone {
+  private static _focusStack: FocusTrapZone[] = []
+  private _root: { current: HTMLElement | null } = { current: null }
+  private _previouslyFocusedElementOutsideTrapZone: HTMLElement
+  private _previouslyFocusedElementInTrapZone?: HTMLElement
+  private windowElement: Window | null
+
+  private createRef = elem => (this._root.current = ReactDOM.findDOMNode(elem) as HTMLElement)
+
+  static propTypes = {
+    as: customPropTypes.as,
+    componentRef: PropTypes.object,
+    className: PropTypes.string,
+    elementToFocusOnDismiss: PropTypes.object,
+    ariaLabelledBy: PropTypes.string,
+    isClickableOutsideFocusTrap: PropTypes.bool,
+    ignoreExternalFocusing: PropTypes.bool,
+    forceFocusInsideTrap: PropTypes.bool,
+    firstFocusableSelector: PropTypes.string,
+    disableFirstFocus: PropTypes.bool,
+    focusPreviouslyFocusedInnerElement: PropTypes.bool,
+  }
+
+  static handledProps = [
+    'as',
+    'className',
+    'componentRef',
+    'elementToFocusOnDismiss',
+    'ariaLabelledBy',
+    'isClickableOutsideFocusTrap',
+    'ignoreExternalFocusing',
+    'forceFocusInsideTrap',
+    'firstFocusableSelector',
+    'disableFirstFocus',
+    'focusPreviouslyFocusedInnerElement',
+  ]
+
+  static defaultProps: IFocusTrapZoneProps = {
+    as: 'div',
+    isClickableOutsideFocusTrap: true,
+  }
+
+  public componentDidMount(): void {
+    FocusTrapZone._focusStack.push(this)
+    const { disableFirstFocus = false } = this.props
+
+    this.windowElement = getWindow(this._root.current)
+    this._previouslyFocusedElementOutsideTrapZone = this._getPreviouslyFocusedElementOutsideTrapZone()
+
+    if (
+      !this._root.current.contains(this._previouslyFocusedElementOutsideTrapZone) &&
+      !disableFirstFocus
+    ) {
+      this.focus()
+    }
+
+    this._hideContentFromAccessibilityTree()
+    this._subscribeToEvents()
+  }
+
+  public render(): JSX.Element {
+    const { className, ariaLabelledBy } = this.props
+    const rest = getUnhandledProps({ handledProps: FocusTrapZone.handledProps }, this.props)
+    const ElementType = getElementType({ defaultProps: FocusTrapZone.defaultProps }, this.props)
+
+    return (
+      <ElementType
+        {...rest}
+        className={className}
+        ref={this.createRef}
+        aria-labelledby={ariaLabelledBy}
+        onKeyDown={this._onKeyboardHandler}
+        onFocusCapture={this._onFocusCapture}
+      >
+        {this.props.children}
+      </ElementType>
+    )
+  }
+
+  public componentWillUnmount(): void {
+    const { ignoreExternalFocusing } = this.props
+
+    FocusTrapZone._focusStack = FocusTrapZone._focusStack.filter((value: FocusTrapZone) => {
```

Comment:
could you, please, suggest which memory issues that will solve? trying to imagine the situation with obsolete reference, but given that `_focusStack` is an instance field it is hard for me to come up with any. Are there any hidden event-based interconnections that could be created?

Reply:
I think nothing specific to memory issues, just removing the unmounted focus trap instance from a stack. For e.g. in a case when not last but FTZ instance in the middle of the stack will get unmounted.

Reply:
Actually, there is no difference of _focusStack usage in fabric-ui and current focusTrapZone.tsx
https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/office-ui-fabric-react/src/components/FocusTrapZone/FocusTrapZone.tsx

I think the reason they've done that way it's to be safe in case ```componentWillUnmount``` will be invoked not for the last instance of FTZ in the stack. So they just filter out that reference and assign the rest to the stack. And nothing more specific than that.